### PR TITLE
Optimize uuid lookup

### PR DESCRIPTION
**Pull request by Jupeyy**
Already wrote that patch when i noticed its not about this, maybe we can use it, else nvm

### CODE REVIEW

File: src/base/tl/sorted_array.h [hunk 1]
```diff
@@ -45,7 +45,7 @@ class sorted_array : public array<T, ALLOCATOR>
 		Function: all
 			Returns a sorted range that contains the whole array.
 	*/
-	range all() { return range(parent::list, parent::list + parent::num_elements); }
+	range all() const { return range(parent::list, parent::list + parent::num_elements); }
```
Comment (heinrich5991) [inline src/base/tl/sorted_array.h:48]: This violates const-safety, you can use the returned range to modify elements in the list.

**Review by heinrich5991 — CHANGES_REQUESTED** [review]
Good idea, code needs at least one fix.

File: src/engine/shared/uuid_manager.cpp [hunk 2]
```diff
@@ -104,12 +109,10 @@ const char *CUuidManager::GetName(int ID) const
 
 int CUuidManager::LookupUuid(CUuid Uuid) const
 {
-	for(int i = 0; i < m_aNames.size(); i++)
+	auto it = ::find_binary(m_aNamesSorted.all(), Uuid);
```
[the remark this answers is not in the document]
Comment (Jupeyy) [reply src/engine/shared/uuid_manager.cpp:112]: Dunno deen already used it in skins xd.

File: src/base/tl/sorted_array.h (hunk 1 above)
Comment (Jupeyy) [reply src/base/tl/sorted_array.h:48]: Yeah thought about that too. But clang with Weverything didn't warn so I don't understand why they allow this

Comment (Jupeyy) [reply src/base/tl/sorted_array.h:48]: I assume that only the pointer variable is const, not the actual variable, e.g. val* const m_ptr

Comment (Jupeyy) [reply src/base/tl/sorted_array.h:48]: But I'll just implement two versions, bcs uuid lookup is const :D

Comment (Jupeyy) [reply src/base/tl/sorted_array.h:48]: https://stackoverflow.com/questions/8788863/can-const-member-function-return-a-non-const-pointer-to-a-data-member#:~:text=5%20Answers&text=A%20const%20member%20function%20can,to%20point%20to%20another%20member).

Only found this in the quick

File: src/engine/shared/uuid_manager.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (Jupeyy) [reply src/engine/shared/uuid_manager.cpp]: i changed it for now, maybe we can think about allowing auto for small for loops, or small functions

**Review by heinrich5991 — APPROVED** [review]
Still violates const-safety, but I don't see a good way to fix it without introducing a `const_range` type.

bors r+

Thanks.

**Comment by Jupeyy** [comment]
how does it?, the pointers arent const, only the members

**Comment by heinrich5991** [comment]
For `std::vector`, the const-ness of the vector is inherited to the data elements (you can see that it doesn't define `begin()` to return a non-const iterator for a constant vector. For `list`, I'd expect the same, but that would require introducing a new iterator/range type for const ranges, probably not worth it.

**Comment by Jupeyy** [comment]
yeah yeah, sry, was just stupid

**Pull request merged**